src/odin/maneuversbuilder.cc

```diff
@@ -3939,5 +3943,36 @@ void ManeuversBuilder::CollapseMergeManeuvers(std::list<Maneuver>& maneuvers) {
   }
 }
 
+void ManeuversBuilder::AddLandmarksFromTripLegToManeuvers(std::list<Maneuver>& maneuvers) {
+  std::vector<LandmarkManeuver> landmarks{};
+  for (auto man = maneuvers.begin(); man != maneuvers.end(); ++man) {
+    // set landmarks correlated with edges in the previous maneuver to the current maneuver
+    if (!landmarks.empty()) {
+      man->set_landmarks(landmarks);
+    }
+    landmarks.clear();
+
+    // accumulate landmarks in the current maneuver
+    double distance_from_begin_to_curr_edge = 0;    // distance from the begin point of the whole
+                                                    // manerver to the begin point of the current edge
+    double maneuver_total_distance = man->length(); // total distance of the maneuver
+
+    for (auto node = man->begin_node_index(); node < man->end_node_index(); ++node) {
+      auto curr_edge = trip_path_->GetCurrEdge(node);
+      if (curr_edge != trip_path_->GetCurrEdge(node + 1)) {
+        // every time we are about to leave an edge, collect all landmarks in it
+        // and reset distance of each landmark to the distance from the landmark to the maneuver point
+        auto curr_landmarks = curr_edge->landmarks();
+        for (auto& l : curr_landmarks) {
+          double new_distance =
+              maneuver_total_distance - l.distance() - distance_from_begin_to_curr_edge;
+          l.set_distance(new_distance);
+        }
+        std::move(curr_landmarks.begin(), curr_landmarks.end(), std::back_inserter(landmarks));
```

Comment:
untested but we need to accumulate the distance along the maneuver to the edge we just looked at
```suggestion
        std::move(curr_landmarks.begin(), curr_landmarks.end(), std::back_inserter(landmarks));
        distance_from_begin_to_curr_edge += curr_edge.length();
```